realm/realm-library/src/objectServer/java/io/realm/SyncSession.java

```diff
@@ -118,6 +124,63 @@ void close() {
         clearScheduledAccessTokenRefresh();
     }
 
+    // This method will be called once all changes have been downloaded.
+    // This method might be called on another thread than the one that called `downloadAllServerChanges`
+    // Be very careful with synchronized blocks.
+    // If the native listener was successfully registered, Object Store guarantee that this method will be called at
+    // least once, even if the session is closed.
+    void notifyAllChangesDownloaded(Long errorcode, String errorMessage) {
+        if (waitingForServerChanges != null) {
+            waitingForServerChanges.handleResult(errorcode, errorMessage);
+        }
+    }
+
+    /**
+     * Calling this method will block until all known remote changes have been download and applied to the Realm.
+     * This will involve network access, so calling this method should only be done from a non-UI thread.
+     * <p>
+     * If the device is offline, this method will return immediately.
+     * <p>
+     * This method cannot be called before the session has been started.
+     *
+     * @throws IllegalStateException if called on the Android main thread.
+     */
+    public void downloadAllServerChanges() {
+        checkMainThread("downloadAllServerChanges() cannot be called from the main thread.");
+
+        // Blocking only happens at the Java layer. To prevent deadlocking the underlying SyncSession we register
+         // an async listener there and let it callback to the Java Session when done. This feel icky at best, but
+         // since all operations on the SyncSession operate under a shared mutex, we would prevent all other actions on the
+         // session, including trying to stop it.
+         // In Java we cannot lock on the Session object either since it will prevent any attempt at modifying the
+         // lifecycle while it is in a waiting state. Thus we use a specialised mutex.
+        synchronized (waitForChangesMutex) {
```

Comment:
Nothing prevents you from doing it right now, since the `downloadAllServerChanges()` method is public. I'm actually a bit sceptical about having the `downloadAllServerChanges()` method public, so feedback on that would be nice as well.